Userland/Libraries/LibTest/Randomized/Generator.h

```diff
@@ -0,0 +1,321 @@
+//                         etc.
+//
+// In case `min == max`, the RandomRun footprint will be smaller, as we'll
+// switch to a `constant` and won't need any randomness to generate that
+// value:
+//
+// Gen::unsigned_int(3,3) -> value 3, RandomRun [] (always)
+//
+// Shrinks towards the smaller argument.
+inline u32 unsigned_int(u32 min, u32 max)
+{
+    VERIFY(max >= min);
+    if (min == max) {
+        return min;
+    }
+    return unsigned_int(max - min) + min;
+}
+
+// Randomly (uniformly) selects a value out of the given arguments.
+//
+// Gen::one_of(20,5,10) --> value 20, RandomRun [0]
+//                      --> value 5,  RandomRun [1]
+//                      --> value 10, RandomRun [2]
+//
+// Shrinks towards the earlier arguments (above, towards 20).
+template<typename... Ts>
+requires(sizeof...(Ts) > 0)
+CommonType<Ts...> one_of(Ts... choices)
+{
+    Vector<CommonType<Ts...>> choices_vec { choices... };
+
+    constexpr size_t count = sizeof...(choices);
+    size_t i = unsigned_int(count - 1);
+    return choices_vec[i];
+}
+
+template<typename T>
+struct Choice {
+    i32 weight;
+    T value;
+};
+
+// Randomly (uniformly) selects a value out of the given weighted arguments.
+//
+// Gen::frequency(
+//   Gen::Choice{5,'x'},
+//   Gen::Choice{1,'o'}
+// )
+//     --> value 'x' (5 out of 6 times), RandomRun [0]
+//     --> value 'o' (1 out of 6 times), RandomRun [1]
+//
+// Shrinks towards the earlier arguments (above, towards 'x').
+template<typename... Ts>
+requires(sizeof...(Ts) > 0)
+CommonType<Ts...> frequency(Choice<Ts>... choices)
+{
+    Vector<Choice<CommonType<Ts...>>> choices_vec { choices... };
+
+    u32 sum = 0;
+    for (auto const& choice : choices_vec) {
+        VERIFY(choice.weight > 0);
+        sum += static_cast<u32>(choice.weight);
+    }
+
+    u32 target = unsigned_int(sum);
+    size_t i = 0;
+    for (auto const& choice : choices_vec) {
+        u32 weight = static_cast<u32>(choice.weight);
+        if (weight >= target) {
+            return choice.value;
+        }
+        target -= weight;
+        ++i;
+    }
+    return choices_vec[i - 1].value;
+}
+
+// An unsigned integer generator in the full u32 range.
+//
+// 8/17 (47%) of the time it will bias towards 8bit numbers,
+// 4/17 (23%) towards 4bit numbers,
+// 2/17 (12%) towards 16bit numbers,
+// 1/17 (6%) towards 32bit numbers,
+// 2/17 (12%) towards edge cases like 0 and NumericLimits::max() of various unsigned int types.
+//
+// Gen::unsigned_int() -> value 3,     RandomRun [0,3]
+//                     -> value 8,     RandomRun [1,8]
+//                     -> value 100,   RandomRun [2,100]
+//                     -> value 5,     RandomRun [3,5]
+//                     -> value 255,   RandomRun [4,1]
+//                     -> value 65535, RandomRun [4,2]
+//                     etc.
+//
+// Shrinks towards 0.
+inline u32 unsigned_int()
+{
+    u32 bits = frequency(
+        // weight, bits
+        Choice { 4, 4 },
```

Comment:
Clang rejects this fwiw. Both the brew clang from llvm@15 for the macOS build, and serenity's patched clang based on 16.06.

Reply:
Looks like this is a clang bug that was fixed sometime b/w clang 16 and clang 17.
https://godbolt.org/z/3r6EqKnha

Reply:
Do I understand correctly that the issue is with a missing `<u32>`? ie. `Choice<u32> { 4, 4 }`